entrance_rando.py

```diff
@@ -0,0 +1,388 @@
+                if all(e.connected_region in er_state.placed_regions for e in lookup):
+                    return False
+
+            # if we're on minimal accessibility and can guarantee the game is beatable,
+            # we can prevent a failure by bypassing future validity checks. this check may be
+            # expensive; fortunately we only have to do it once
+            if perform_validity_check and world.options.accessibility == Accessibility.option_minimal \
+                    and world.multiworld.has_beaten_game(er_state.collection_state, world.player):
+                # ensure that we have enough locations to place our progression
+                accessible_location_count = 0
+                prog_item_count = len(er_state.collection_state.prog_items[world.player])
+                # short-circuit location checking in this case
+                if prog_item_count == 0:
+                    return True
+                for region in er_state.placed_regions:
+                    for loc in region.locations:
+                        if loc.can_reach(er_state.collection_state):
+                            accessible_location_count += 1
+                            if accessible_location_count >= prog_item_count:
+                                perform_validity_check = False
+                                # pretend that this was successful to retry the current stage
+                                return True
+
+            unplaced_entrances = [entrance for region in world.multiworld.get_regions(world.player)
+                                  for entrance in region.entrances if not entrance.parent_region]
+            unplaced_exits = [exit_ for region in world.multiworld.get_regions(world.player)
+                              for exit_ in region.exits if not exit_.connected_region]
+            entrance_kind = "dead ends" if dead_end else "non-dead ends"
+            region_access_requirement = "requires" if require_new_regions else "does not require"
+            raise EntranceRandomizationError(
+                f"None of the available entrances are valid targets for the available exits.\n"
+                f"Randomization stage is placing {entrance_kind} and {region_access_requirement} "
+                f"new region access by default\n"
+                f"Placeable entrances: {lookup}\n"
+                f"Placeable exits: {placeable_exits}\n"
+                f"All unplaced entrances: {unplaced_entrances}\n"
+                f"All unplaced exits: {unplaced_exits}")
+
+    er_targets = [entrance for region in world.multiworld.get_regions(world.player)
+                  for entrance in region.entrances if not entrance.parent_region]
+    exits = [ex for region in world.multiworld.get_regions(world.player)
+             for ex in region.exits if not ex.connected_region]
+    if len(er_targets) != len(exits):
+        raise EntranceRandomizationError(f"Unable to randomize entrances due to a mismatched count of "
```

Comment:
Oh this is interesting

I feel like "I'm gonna make a bunch of extra entries that don't need to be connected to" could be a valid use case, e.g. when "randomizing warp song destinations" where you want the 6 warp songs to go to completely random regions

Again, could just be something the world would do itself though.

Reply:
Yeah that doesn't seem like a GER problem to me. Having a matched number of "somewhere to go from" and "somewhere to go" seems pretty fundamentally important on a conceptual level, in addition to a technical one

Reply:
In a case like that, the world dev could choose which of those 6 random regions to use in advance

Reply:
I'll be solving this problem with The Witness soon, maybe it's something you're interested in:

I'll basically have a directed region graph where not everything can be reached from the starting region, and I want to be able to pick warps (connections from Entry to random regions) **such that** everything is reachable.

I'm ok with it not being a GER problem, but it's an ER problen and a fun graphing problem, so we'll see how I'll solve it :)

Anyway, consider this resolved if you want to

Reply:
Yeah within the scope of GER, for a case like this, what I would immediately think to do is assign groups for the "islands" (possibly psuedo-randomly if desired) and have ER do grouping within the island subgroups. Then you can pre-place warps from start to "somewhere in each island" and GER would solve the rest